### DISCUSSION TITLE
Feature request - Show sent message before it returned from firebase?

### DISCUSSION THREAD

**Discussion by itaishalom** [flyerhq/flutter_chat_ui#118] [Q&A] [answered]
Hi, I'm using the firebase_core also but I changed the backend to another server that is communicating with firebase (due to security reasons).
There is a latency, and it causes written message to disappear from the input widget and show again at the chat after 3-4 seconds. Is there an easy way to add the message to the list of messages prior getting if back from the server?

**Comment by demchenkoalex** [comment]
Hey @itaishalom I don't think this is possible. Of course you're the one who supplies the messages array to the chat UI, and you can easily add a message to the array, and it will be shown, however, when the message will be registered on a backend, it will 100% have a different ID and you will need to find your temporarily added message and update it's ID and I am afraid ID change will also trigger animation so you will get something like, message appears, disappears, appears. I don't remember if I have an option to disable the animation, I think it can be added, but it will not resolve initial issue that you will be adding 2 messages for one, so you have problems with queues and stuff and this is just too complex topic to start off, we are talking about offline first etc. Maybe if your backend would be able to accept temporarily created IDs and use them as a server side IDs, and then make a change in the chat UI to compare only IDs, something like that can work, but it is definitely not doable with Firebase, or at least too complex to even spend time on it.

**Comment by demchenkoalex** [comment]
After talking to @cristaloleg we think that having 2 IDs might help, like having current id as an indication of a message added to the chat but not received on the server and `messageServerId` or something which will be set after message is received on the backend. In this case there will be no delete/insert animations, it will be just a local `messageServerId` update. And you will use second id for all business logic and endpoints, and first id (`id`) as an indication of message added to the chat, like offline mode. That's just an idea and not verified ATM.

**Reply by itaishalom** [reply]
That's sounds like a good start. Let me know if you want  help implementing it.

**Reply by demchenkoalex** [reply]
Just so we are on the same page, we are talking about adding 1 field to types and basically that's it, everything else about filling those etc etc is on developer's side, at least how I see that ATM.

**Reply by itaishalom** [reply]
Yes so i think that what you suggested could work, just not sure about how to disable the animations but I'll be happy to figure it out. Do you think you could do it?

**Reply by demchenkoalex** [reply]
I can take a look on weekend :)

**Reply by itaishalom** [reply]
Cool, thank you!

**Reply by demchenkoalex** [reply]
Added the `remoteId` property to all messages in [v1.4.4](https://github.com/flyerhq/flutter_chat_ui/releases/tag/v1.4.4). Also added it to `copyWith` function. Setting it will not trigger the animation. According to our ideas, `id` will be used as a "local" id, and `remoteId` will be set when backend will return success (it will be an ID that server generates), and for all additional operations, like update and delete, we will need to use `remoteId`. Didn't verify these thoughts yet.

**Comment by itaishalom** [comment, marked as answer]
Awesome! I'll test it on Thursday 😉

**Reply by itaishalom** [reply]
What should I do to make it work? Any changes are required to do on my code?

**Reply by demchenkoalex** [reply]
Please read all the conversation from the beginning :) I can't tell you exactly what to do, I don't know how it should work, all our messages were theoretical, now you can try do implement them on practice :) from the library point of view you should have everything ready, everything else is a developer.

**Reply by itaishalom** [reply]
Ok so just to be cleared, this should be implemented on my project or on the library and PR you?

**Reply by demchenkoalex** [reply]
In the project :) library should provide everything needed according to our "theoretical" conversation

**Reply by itaishalom** [reply]
Got you, thanks.

**Reply by itaishalom** [reply]
I ended up with another solution.
I modified you firebase core example into two parts - 
First - Get all messages currently on the room.
Second - New listener that listens to new updates from **now on**.
If the new message arrived my my Id, don't add it, else add it to the list.
Notify UI with the new messages.
